### DISCUSSION TITLE
Image flashes when screen is recomposed

### DISCUSSION THREAD

**Discussion by DrifterAtSea** [coil-kt/coil#882] [Q&A]
I am using animations to show and hide a screen with Jetpack Compose. I have a screen that slides in from the right when it is made visible and then slide it back to the right when the screen is hidden.

Jetpack Compose animations recompose their content both when making the composable visible and invisible.

On the screen that is being animated, I have an Image and use rememberImagePainter and crossfade is set to false.

When the screen is made visible, the image is loaded from the network and is displayed as expected. But when the screen is made invisible by animating off to the right, the image will temporarily flash as the screen is being recomposed.

It is difficult to say what the problem is. It is clear that Jetpack Compose is recomposig the screen and therefore rememberImagePainter is being called again. But why the flashing? If Coil already has loaded the image, why is it not just reloaded from cache without the flash?

**Comment by tealang7012** [comment]
meat the same problem. wish any response.

**Reply by DrifterAtSea** [reply]
I developed a solution to fix image flashing. It's called "Image Manager" and is part of Jetmagic - another framework I developed for Compose:

[https://github.com/JohannBlake/Jetmagic](https://github.com/JohannBlake/Jetmagic)

**Reply by colinrtwhite** [reply]
This is likely an issue with how `ImagePainter` emits an `Empty` state before the `Loading` state. I'm working on a fix at the moment (though I can't offer an eta unfortunately).

**Reply by omar-leyva** [reply]
Curious if there is any updates on this? I'm seeing this issue when switching between compose screens using a `TabRow`

**Reply by linean** [reply]
I've noticed the same issue during re-composition. Fix would be really appreciated :pray: Maybe it's possible to check local cache before emitting empty state?

**Reply by colinrtwhite** [reply]
Yep, this should be fixed in 2.0.0-alpha04 and newer.

**Reply by TokenYc** [reply]
> Yep, this should be fixed in 2.0.0-alpha04 and newer.
it's better but still a little flash when i use 2.0.0-alpha05.
i have recorded the video: https://wwe.lanzouo.com/ivKWyx50tgd
if i use local image with painterResource,there is no flash.

**Reply by colinrtwhite** [reply]
@TokenYc Yep I'm able to reproduce this in a few instances in the Jetpack Compose samples repo. I think this is due to an issue with `BoxWithConstraints` as we're drawing the image on the first frame, but it's not being displayed. Hopefully I can find a way to work around it (or maybe it needs to be fixed inside Jetpack Compose).

Either way it should return images from the memory cache synchronously significantly more often in 2.0.0-alpha04.

**Reply by linean** [reply]
I've tried `alpha05` and issue is still there. It may be related with non standard usage on my side (RecyclerView + ComposeView).
Here you can find sample: https://github.com/linean/coil-compose-glitch

**Reply by svenjacobs** [reply]
In my case the flashing of images is gone with `2.0.0-alpha05`. It occurred when showing/hiding a `ModalBottomSheetLayout`.

**Update:** I have to partially withdraw my statement. I see the flashing when `interceptorDispatcher(Dispatchers.IO)` is configured on `ImageLoader.Builder`. Without that option there is no flashing.

**Reply by colinrtwhite** [reply]
@svenjacobs That's intended behaviour Setting `interceptorDispatcher(Dispatchers.IO)` moves all work to an IO thread - including the memory cache check.

I'm hoping https://github.com/coil-kt/coil/pull/1048 should resolve most synchronous cache issues. I'm still seeing occasional white flashes when using `BoxWithConstraints`, though that seems like an issue in Jetpack Compose and is not possible to solve in Coil.

**Reply by RobertMenke** [reply]
I'm getting this behavior consistently when I use a local drawable (ex: `rememberAsyncImagePainter(ContextCompat.getDrawable(context, R.drawable.clock_icon_white_clipped))`), but when I use a `File` I don't experience image flickering.

This seems to happen whether or not I'm using `BoxWithConstraints`. I'm on 2.0.0-rc01. Happy to invite you to the repo (private) if you want to take a look. 

Appreciate all the hard work. Coil is absolutely fantastic ❤️ .

**Reply by RobertMenke** [reply]
My apologies, I actually just fixed this by switching from using `ContextCompat.getDrawable` to just providing the resource id.

Converting this
```kotlin
rememberAsyncImagePainter(ContextCompat.getDrawable(context, R.drawable.clock_icon_white_clipped))
```

into this
```kotlin
rememberAsyncImagePainter(R.drawable.clock_icon_white_clipped)
```

Fixed the flickering in my case

**Reply by mainrs** [reply]
The flicking solution you mentioned above does not work for Compose Desktop and coil3. I have the same issues there
